cmd/tsbs_load_clickhouse/creator.go

```diff
@@ -0,0 +1,238 @@
+	// nginx,accepts,active,handled,reading,requests,waiting,writing
+
+	i := 0
+	for {
+		var err error
+		var line string
+
+		if i == 0 {
+			// read first line - list of tags
+			d.tags, err = br.ReadString('\n')
+			if err != nil {
+				fatal("input has wrong header format: %v", err)
+			}
+			d.tags = strings.TrimSpace(d.tags)
+		} else {
+			// read the second and further lines - metrics descriptions
+			line, err = br.ReadString('\n')
+			if err != nil {
+				fatal("input has wrong header format: %v", err)
+			}
+			line = strings.TrimSpace(line)
+			if len(line) == 0 {
+				// empty line - end of header
+				break
+			}
+			// append new table/columns set to the list of tables/columns set
+			d.cols = append(d.cols, line)
+		}
+		i++
+	}
+}
+
+// loader.DBCreator interface implementation
+func (d *dbCreator) DBExists(dbName string) bool {
+	db := sqlx.MustConnect(dbType, getConnectString(false))
+	defer db.Close()
+
+	sql := "SELECT name, engine FROM system.databases WHERE name = '" + dbName + "'"
+	if debug > 0 {
+		fmt.Printf(sql)
+	}
+	var rows []struct {
+		Name   string `db:"name"`
+		Engine string `db:"engine"`
+	}
+
+	err := db.Select(&rows, sql)
+	if err != nil {
+		panic(err)
+	}
+	for _, row := range rows {
+		if row.Name == dbName {
+			return true
+		}
+	}
+
+	return false
+}
+
+// loader.DBCreator interface implementation
+func (d *dbCreator) RemoveOldDB(dbName string) error {
+	db := sqlx.MustConnect(dbType, getConnectString(false))
+	defer db.Close()
+
+	sql := "DROP DATABASE IF EXISTS " + dbName
+	if _, err := db.Exec(sql); err != nil {
+		panic(err)
+	}
+	return nil
+}
+
+// loader.DBCreator interface implementation
+func (d *dbCreator) CreateDB(dbName string) error {
+	// Connecto to ClickHouse in general and CREATE DATABASE
+	db := sqlx.MustConnect(dbType, getConnectString(false))
+	sql := "CREATE DATABASE " + dbName
+	_, err := db.Exec(sql)
+	if err != nil {
+		panic(err)
+	}
+	db.Close()
+	db = nil
+
+	// Connect to specified database within ClickHouse
+	db = sqlx.MustConnect(dbType, getConnectString(true))
+	defer db.Close()
+
+	// d.tags content:
+	//tags,hostname,region,datacenter,rack,os,arch,team,service,service_version,service_environment
+	//
+	// Parts would contain
+	// 0: tags - reserved word - tags mark
+	// 1:
+	// N: actual tags
+	// so we'll use tags[1:] for tags specification
+	parts := strings.Split(strings.TrimSpace(d.tags), ",")
+	if parts[0] != "tags" {
+		return fmt.Errorf("input header in wrong format. got '%s', expected 'tags'", parts[0])
+	}
+	createTagsTable(db, parts[1:])
+	tableCols["tags"] = parts[1:]
+
+	// d.cols content are lines (metrics descriptions) as:
+	//cpu,usage_user,usage_system,usage_idle,usage_nice,usage_iowait,usage_irq,usage_softirq,usage_steal,usage_guest,usage_guest_nice
+	//disk,total,free,used,used_percent,inodes_total,inodes_free,inodes_used
+	//nginx,accepts,active,handled,reading,requests,waiting,writing
+	for _, cols := range d.cols {
```

Comment:
You have a `createTagsTable` function above for the tags table, why not extract a similar one for the metrics tables instead of having all of the logic in this function? Would make this function a bit easier to read (and test).

Reply:
This makes sense

Reply:
Updated